test/integration/metrics/metrics_test.go

```diff
@@ -111,3 +114,130 @@ func TestApiserverMetrics(t *testing.T) {
+			},
+		}
+	}
+
+	callOrDie(c.Create(context.TODO(), makePod("foo"), metav1.CreateOptions{}))
+	appendExpectedMetric(metricLabels("", "v1", "pods", "", "resource", "POST"))
+	callOrDie(c.Update(context.TODO(), makePod("bar"), metav1.UpdateOptions{}))
+	appendExpectedMetric(metricLabels("", "v1", "pods", "", "resource", "PUT"))
+	callOrDie(c.UpdateStatus(context.TODO(), makePod("bar"), metav1.UpdateOptions{}))
+	appendExpectedMetric(metricLabels("", "v1", "pods", "status", "resource", "PUT"))
+	callOrDie(c.Get(context.TODO(), "foo", metav1.GetOptions{}))
+	appendExpectedMetric(metricLabels("", "v1", "pods", "", "resource", "GET"))
+	callOrDie(c.List(context.TODO(), metav1.ListOptions{}))
+	appendExpectedMetric(metricLabels("", "v1", "pods", "", "namespace", "LIST"))
+	callOrDie(nil, c.Delete(context.TODO(), "foo", metav1.DeleteOptions{}))
+	appendExpectedMetric(metricLabels("", "v1", "pods", "", "resource", "DELETE"))
+	// cluster-scoped LIST of namespace-scoped resources
+	callOrDie(client.CoreV1().Pods(metav1.NamespaceAll).List(context.TODO(), metav1.ListOptions{}))
+	appendExpectedMetric(metricLabels("", "v1", "pods", "", "cluster", "LIST"))
+
+	// Cluster-scoped resource
+	cn := client.CoreV1().Namespaces()
+	makeNamespace := func(labelValue string) *v1.Namespace {
+		return &v1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   "foo",
+				Labels: map[string]string{"foo": labelValue},
+			},
+		}
+	}
+
+	callOrDie(cn.Create(context.TODO(), makeNamespace("foo"), metav1.CreateOptions{}))
+	appendExpectedMetric(metricLabels("", "v1", "namespaces", "", "resource", "POST"))
+	callOrDie(cn.Update(context.TODO(), makeNamespace("bar"), metav1.UpdateOptions{}))
+	appendExpectedMetric(metricLabels("", "v1", "namespaces", "", "resource", "PUT"))
+	callOrDie(cn.UpdateStatus(context.TODO(), makeNamespace("bar"), metav1.UpdateOptions{}))
+	appendExpectedMetric(metricLabels("", "v1", "namespaces", "status", "resource", "PUT"))
+	callOrDie(cn.Get(context.TODO(), "foo", metav1.GetOptions{}))
+	appendExpectedMetric(metricLabels("", "v1", "namespaces", "", "resource", "GET"))
+	callOrDie(cn.List(context.TODO(), metav1.ListOptions{}))
+	appendExpectedMetric(metricLabels("", "v1", "namespaces", "", "cluster", "LIST"))
+	callOrDie(nil, cn.Delete(context.TODO(), "foo", metav1.DeleteOptions{}))
+	appendExpectedMetric(metricLabels("", "v1", "namespaces", "", "resource", "DELETE"))
+
+	// Verify if all metrics were properly exported.
+	metrics, err := scrapeMetrics(s)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	samples, ok := metrics["apiserver_request_total"]
+	if !ok {
+		t.Fatalf("apiserver_request_total metric not exposed")
+	}
+
+	hasLabels := func(current, expected model.Metric) bool {
+		for key, value := range expected {
+			if current[key] != value {
+				return false
+			}
+		}
+		return true
+	}
+
+	for _, expectedMetric := range expectedMetrics {
+		found := false
+		for _, sample := range samples {
+			if hasLabels(sample.Metric, expectedMetric) {
```

Comment:
I think this methodology won't detect if two requests have opposite bugs, i.e. some aspect of their metrics are flipped?

Reply:
Opened https://github.com/kubernetes/kubernetes/issues/95636